### PR TITLE
Add help for new developers

### LINKED ISSUE #293: Add developer documentation
We should add some developer docs covering:

- How to setup a development environment
- How to run the tests
- How to build the docs
- A loose outline of the code structure

### PR DESCRIPTION
**Pull request by davfsa**
- Add code of conduct and contributing guidelines
- Improve and organize gitignore
- Add codespell and isort to pre-commit
- Add pull request checklist

---

Hope this helps out. I probably missed some info, so please point it out and I will append it!

Closes #293

### CODE REVIEW

**Comment by davfsa** [comment]
Just realised, forgot to add documentation on how to build the docs. Will do later when I get time!

**Comment by jcrist** [comment]
Thanks for the PR! When I wrote up issue #293 I was thinking this was work *I* would need to do to make it easier for you to start contributing, not work that *you* should do. Was definitely not expecting this :).

If you wanted to take a pass at adding notes on building the docs that'd be welcome. After that I'll want to make some edits myself (I don't agree with 100% of everything you've written). But in general this is a nice improvement, thanks for the help!

**Comment by davfsa** [comment]
While setting up the project in my local machine, I found it that it would maybe be nice to annotate everything I found and PR it :)

The pull request is open for you to make commits to if you want, sorry for any incorrect information I may have put :sweat_smile:

**Comment by davfsa** [comment]
Added a section on how to build the documentation and fixed some other issues I found :)

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -27,22 +27,18 @@ jobs:
...
+      - name: Run pre-commit hooks
         uses: pre-commit/action@v3.0.0
 
       - name: mypy
-        run: py.test tests/test_mypy.py
+        run: pytest tests/test_mypy.py
 
       - name: pyright
-        run: py.test tests/test_pyright.py
+        run: pytest tests/test_pyright.py
```
Comment (jcrist) [inline .github/workflows/ci.yml:41]: `py.test` is an alternative older spelling of the CLI, my muscle memory still defaults to it over `pytest`. Either is fine.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,12 +1,24 @@
...
-    - id: flake8
-      language_version: python3
+      - id: flake8
+        language_version: python3
+
+  - repo: https://github.com/codespell-project/codespell
+    rev: v2.2.2
+    hooks:
+      - id: codespell
+        language_version: python3
```
Comment (jcrist) [inline .pre-commit-config.yaml:24]: I've never used `codespell` before, but it seems nice. It did find a bunch of false-positives in the code files, for now I've excluded all code files so it just runs on our docs.

Note that when adding new pre-commit hooks you'll need to run them manually on every file to ensure things pass. I did this using `pre-commit run --all-files` and fixed the issues that popped up.

File: .github/CODE_OF_CONDUCT.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
+# Code of Conduct
```
Comment (jcrist) [inline .github/CODE_OF_CONDUCT.md:1]: IMO adding a code of conduct is something that should really only be done by project maintainers, and isn't something that should be taken lightly. I'm fine with using the Contributor Covenant and will merge this as is, but I would have preferred to see some discussion first before submitting such a change.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -27,22 +27,18 @@ jobs:
         with:
           python-version: "3.10"
 
-      - name: Install Dependencies
+      - name: Build msgspec and install dependencies
         run: |
-          pip install pytest mypy pyright msgpack pyyaml tomli tomli_w coverage
+          pip install coverage -e ".[test]"
```
Comment (jcrist) [inline .github/workflows/ci.yml:32]: I moved all the test and docs dependencies into extras in the `setup.py`, which should hopefully make it easier to set up a local development environment.

You can now run

```
pip install -e ".[dev]"
```

to get everything required for developing tests & docs.

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
+# How to Contribute
```
Comment (jcrist) [inline .github/CONTRIBUTING.md:1]: Thanks for the original pass at this @davsfa! I simplified the guidelines here a bit (and expanded them in other places). By its nature `msgspec` will never be a good project for Python beginners to contribute to - as such I dropped explanations that would only be useful for those new at Python and OSS development.

Following the ["write for one person"](https://wizardzines.com/comics/write-for-one-person/) idea, this document is targeted at someone who has contributed on GitHub before, and has some experience with Python development workflows.

**Review by jcrist — APPROVED** [review]
Thanks @davsfa! I pushed up some fixes/edits. Since I amended your commit it's hard to compare my changes with what you originally had, but I left a few comments of explanation for major changes. Thanks for the work here, hopefully this eases the onboarding experience for new contributors :).

**Pull request merged**